### PR TITLE
[21] Add OneForOne restart logic

### PR DESCRIPTION
**Pull request by roman**
### Development Notes

* Now there is an specific notification for start failures
* The event failure reporter is way nicer now

### Acceptance Criteria

* [x] Goroutines get restarted on failure every time they fail
* [x] Unit tests

### CODE REVIEW

File: c/types.go [hunk 1]
```diff
@@ -82,27 +100,66 @@ type NotifyStartFn = func(startError)
 // with the supervisor's ChildSpec.
 type ChildSpec struct {
 	name     string
+	tag      ChildTag
```
Comment (julio-garcia-fc) [inline c/types.go]: It's been a while, so I have forgotten some things, but I am a bit confused. The comment above leads me to believe that `ChildSpec` is, by definition, a worker. Can you clarify? Maybe the abstraction is getting in the way, and explicit WorkerChildSpec and SupervisorChildSpec might be better than the type? I am really not sure. Just thinking as I type.

File: c/core.go [hunk 2]
```diff
@@ -152,6 +159,22 @@ func waitTimeout(
...
+	prevChild Child,
+	parentName string,
+	notifyCh chan<- ChildNotification,
+) (Child, error) {
+	newChild, err := cs.Start(parentName, notifyCh)
+	if err != nil {
+		return Child{}, err
+	}
+	// TODO: When working on treshold restart, verify the create date of
+	// prevChild; if the spec treshold has passed, restart the count here
```
Comment (julio-garcia-fc) [inline c/core.go]: Implying that it will be treated as a brand new child?

File: c/types.go [hunk 3]
```diff
@@ -82,27 +100,66 @@ type NotifyStartFn = func(startError)
...
+func (cs ChildSpec) Tag() ChildTag {
+	return cs.tag
+}
+
+// IsWorker indicates if this child is a worker
+func (cs ChildSpec) IsWorker() bool {
+	return cs.tag == Worker
+}
+
 // Child is the runtime representation of an Spec
```
Comment (julio-garcia-fc) [inline c/types.go]: s/an/a/

File: c/types.go [hunk 4]
```diff
@@ -82,27 +100,66 @@ type NotifyStartFn = func(startError)
...
+
+// IsWorker indicates if this child is a worker
+func (c Child) IsWorker() bool {
+	return c.spec.IsWorker()
 }
 
 // ChildNotification reports when a child has terminated; if it terminated with
 // an error, it is set in the err field, otherwise, err will be nil.
 type ChildNotification struct {
+	name        string
```
Comment (julio-garcia-fc) [inline c/types.go]: Why not also the tag (sorry, I am forgetting stuff, which may indicate the need to have a picture of the model 🤷‍♂ )

File: s/children.go [hunk 5]
```diff
@@ -42,14 +42,21 @@ func stopChildren(
...
+		if terminationErr != nil {
+			// if a child fails to stop (either because of a legit failure or a
+			// timeout), we store the terminationError so that we can report all of them
+			// later
+			childErrMap[cs.Name()] = terminationErr
+
+			// we also notify that the process failed
+			eventNotifier.ProcessFailed(ch.RuntimeName(), terminationErr)
+		} else {
+			// we need to modify that the process stopped
```
Comment (julio-garcia-fc) [inline s/children.go]: s/modify/notify/

File: s/children.go [hunk 6]
```diff
@@ -69,23 +76,34 @@ func startChildren(
...
 		// NOTE: The error handling code bellow gets executed when the children
 		// fails at start time
-		if err != nil {
-			cRuntimeName := strings.Join([]string{runtimeName, cs.Name()}, "/")
-			eventNotifier.ProcessStopped(cRuntimeName, startTime, err)
+		if startErr != nil {
+			cRuntimeName := strings.Join([]string{runtimeName, cs.Name()}, childSepToken)
+			// NOTE: we only notify when child is a worker because sub-trees
+			// supervisors are responsible of their own notification
+			eventNotifier.ProcessStartFailed(cRuntimeName, startErr)
```
Comment (julio-garcia-fc) [inline s/children.go]: The comment above is out of date? The worker check is commented out.

File: s/one_for_one_test.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+		context.TODO(),
+		parentName,
+		[]s.Opt{
+			s.WithChildren(child1),
+		},
+		func(em EventManager) {
+			// NOTE: we won't stop the supervisor until the child has failed at least
+			// once
+			evIt := em.Iterator()
+			// 1) Wait till all the three is up
```
Comment (julio-garcia-fc) [inline s/one_for_one_test.go]: You mean the one?

File: s/one_for_one_test.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+	events, err := ObserveSupervisor(
+		context.TODO(),
+		parentName,
+		[]s.Opt{s.WithSubtree(tree1)},
+		func(em EventManager) {
+			// NOTE: we won't stop the supervisor until the child has failed at least
+			// once
+			evIt := em.Iterator()
+			// 1) Wait till all the three is up
```
Comment (julio-garcia-fc) [inline s/one_for_one_test.go]: Wrong comment again?

File: s/one_for_one_test.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+		},
+	)
+
+	assert.NoError(t, err)
+
+	AssertExactMatch(t, events,
+		[]EventP{
+			// start children from left to right
+			ProcessStarted("root/child1"),
+			ProcessStarted("root"),
```
Comment (julio-garcia-fc) [inline s/one_for_one_test.go]: This is always counter intuitive to me. I would always expect the root to precede the child (prefix). Is this more like infix. prefix or postfix notation?

File: s/one_for_one_test.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+	assert.NoError(t, err)
+
+	AssertExactMatch(t, events,
+		[]EventP{
+			// start children from left to right
+			ProcessStarted("root/subtree1/child1"),
+			ProcessStarted("root/subtree1"),
+			ProcessStarted("root"),
+			// ^^^ 1) Wait till root starts
+			ProcessFailed("root/subtree1/child1"),
```
Comment (julio-garcia-fc) [inline s/one_for_one_test.go]: would `subtrree1` also have been marked as failed?

File: s/one_for_one_test.go [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+			// ^^^ 2) We see the startFailingChild1 causing the error
+			ProcessStarted("root/subtree1/child1"),
+			// ^^^ 3) After 1st (re)start we stop
+			ProcessStopped("root/subtree1/child1"),
+			ProcessStopped("root/subtree1"),
+			ProcessStopped("root"),
```
Comment (julio-garcia-fc) [inline s/one_for_one_test.go]: I would expect the order of the above to be reversed. It may not make any difference for the test, but intuitively, I would expect it to be the opposite of how you check the start above.

File: s/start.go [hunk 12]
```diff
@@ -67,19 +102,40 @@ func runMonitorLoop(
...
 			return nil
-		case /* childNotification = */ <-notifyCh:
-			// TODO: Deal with errors on children
+		case childNotification := <-notifyCh:
+			oldChild, ok := children[childNotification.Name()]
+
+			if !ok {
+				// TODO: Expand on this case, I think this is highly unlikely, but would
+				// like to exercise this branch in test somehow (if possible)
+				panic("something horribly wrong happened here")
```
Comment (julio-garcia-fc) [inline s/start.go]: Note the `name`, `tag`, ...?

File: c/core.go (hunk 2 above)
Comment (roman) [reply c/core.go]: Say we have a maximum error tolerance of "3 restarts in a 10 second window", and a 4th error happened on second 11, we restart the count here.

Every time a Child restarts, a new `ProcessCreated` event is triggered.

File: c/types.go (hunk 1 above)
Comment (roman) [reply c/types.go]: In practice, this is correct, you see that a sub-tree supervisor is implemented in terms of a [child/worker goroutine](https://github.com/capatazlib/go-capataz/blob/master/s/subtree.go#L32).

I've thought about this, IMO, adding a _type heriarchy_ of `WorkerChildSpec` and `SupervisorChildSpec` enforces us to define an interface implementation for each sub-type (I foresee only two of them), that is similar in all senses except an if statement for sending notifications to the events channel.

To summarize: a `Child` is either be a sub-tree (`Supervisor`) or a leaf (`Worker`), the only difference right now is that we deal with notifications differently [when the `Child` is a `Worker`](https://github.com/capatazlib/go-capataz/pull/26/files#diff-75e644655432015c82ca534acf607b3cR103) and when the `Child` is a `Supervisor` (supervisors are responsible on their own notifications).

File: c/types.go (hunk 4 above)
Comment (roman) [reply c/types.go]: Yep, that works, it doesn't hurt having it in the Events emitted

File: s/children.go (hunk 6 above)
Comment (roman) [reply s/children.go]: Yes, that is outdated, thanks for catching this, removing the comment

File: s/one_for_one_test.go (hunk 7 above)
Comment (roman) [reply s/one_for_one_test.go]: s/three/tree/

File: s/one_for_one_test.go (hunk 9 above)
Comment (roman) [reply s/one_for_one_test.go]: Leafs of the tree are the workers of the system. They start first and once all of them start, the supervisor is considered started (monitoring their errors). The start order will be pre-order (`LeftToRight`) or post-order (`RightToLeft`) depending on the `SupervisorSpec` settings.

We do traverse the root first, but the parents are the last bit that start in the system. If there is an error on any started worker before the supervisor starts, the worker child will block in a write chan and the supervisor will pick it up once it starts.

File: s/one_for_one_test.go (hunk 8 above)
Comment (roman) [reply s/one_for_one_test.go]: s/three/tree/

File: s/one_for_one_test.go (hunk 10 above)
Comment (roman) [reply s/one_for_one_test.go]: No, the `subtree1` is a supervisor, and it will fail only if the error threshold of the child is surpassed. In this test-case we are not defining the explicit error tolerance of the `subtree1` supervisor just yet -- we are tackling that implementation on a different PR. For now, assume they can fail infinitely and they get recovered by the `subtree1` supervisor every time.

File: s/one_for_one_test.go (hunk 9 above)
Comment (roman) [reply s/one_for_one_test.go]: I'm going to write the lines above in the comments so that expectations of the tests becomes very explicit and sense can be built.

File: s/one_for_one_test.go (hunk 11 above)
Comment (roman) [reply s/one_for_one_test.go]: I've the impression this is related to the [counter-intuitive order comment](https://github.com/capatazlib/go-capataz/pull/26/files#r392700332) you mentioned in this PR.

File: s/start.go (hunk 12 above)
Comment (roman) [reply s/start.go]: Good idea

**Comment by roman** [comment]
> Left a few comments. Some may reflect the fact that I have forgotten a few things. Sorry.

No worries, I've forgotten a lot of the code, you see me renaming some of the variables to be more explicit to bring everything back.

File: s/one_for_one_test.go (hunk 10 above)
Comment (julio-garcia-fc) [reply s/one_for_one_test.go]: Got it!

File: s/one_for_one_test.go (hunk 9 above)
Comment (julio-garcia-fc) [reply s/one_for_one_test.go]: I see, and you have explained this before. Sorry. What you explain is not wrong, but there may be overloaded concepts, perhaps (???). What is counter intuitive may be this: suppose the the system is up and running (everything marked as started and running smoothly). If a child dies, completely dead (restarts fail, etc), what happens to the root supervisor: marked as in error only, but still running? Is error state equivalent as not running? What if all the other children are still running? Is there a distinction between a running supervisor and its children running? Is there a distinction between the error state of a supervisor and the error state of the children? I am not at all advocating more complexity or any changes. I am just wondering if more clarity in the comments or README would be helpful. I see you have done some of that already and it is great.

File: stest/README.md [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+* `WorkerFailed`
+* `SupervisorStopped`
+* `WorkerStopped`
+
+Check the test implementation to see how they are used.
+
+#### A note about event order in assertions
+
+Some of the readers may find counter-intuitive the order of the collected
+events. The supervision system will spawn goroutines the leafs of the tree
```
Comment (julio-garcia-fc) [inline stest/README.md]: s/will spawn goroutines the leafs/will spawn goroutines for the leafs/

Comment (julio-garcia-fc) [reply stest/README.md]: I think this is what is throwing me off: `supervision system` != `supervisor (root or otherwise)`. The only role of the `supervisor` is to _**monitor**_ a _running_ `worker`. Maybe a better name for the `supervision system` would help? Something like `environment monitor`, `supervision manager`, `<something> manager`, `management kernel`, `kernel`, ... Again, just pondering.
